### PR TITLE
Issue #2137: Add support json variable to web servicetask

### PR DESCRIPTION
**Pull request by cdeneux**
Please find here a contribution about issue #2137. See also the following discussion in the forum: https://forum.flowable.org/t/dataobject-vs-custom-types/2166

#### Check List:
* Unit tests:
  * Unit tests added about Json variable to XML part (SOAP request) conversion,
  * Unit tests added about XML part (of SOAP response) to JSon variable conversion
* Documentation **must be updated** about:
  * Converting JSon variable into XML part of the SOAP request,
  * Converting an XML part of the SOAP response as JSon data object

### CODE REVIEW

File: modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/Operands.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package org.flowable.engine.impl.webservice;
```
Comment (dbmalkovsky) [inline modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/Operands.java]: Missing required license header

Comment (cdeneux) [reply modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/Operands.java]: Thanks, I will add it

File: modules/flowable-bpmn-model/src/main/java/org/flowable/bpmn/model/JsonDataObject.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+public class JsonDataObject extends ValuedDataObject {
+
+    @Override
+    public void setValue(Object value) {
+    	if (value instanceof String && !StringUtils.isEmpty(((String) value).trim())) {
+    		try {
+				this.value = DateFormat.getDateTimeInstance().parse((String) value);
+			} catch (ParseException e) {
+				System.out.println("Error parsing Date string: " + value);
```
Comment (dbmalkovsky) [inline modules/flowable-bpmn-model/src/main/java/org/flowable/bpmn/model/JsonDataObject.java]: Why is this a `println` and not a `throw`?  As it is the `this.value` has no value....

File: modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java [hunk 3]
```diff
@@ -138,4 +142,171 @@ public void testWebServiceInvocationWithEndpointAddressConfigured() throws Excep
...
+
+        processEngineConfiguration.addWsEndpointAddress(
+                new QName("http://webservice.impl.engine.flowable.org/", "CounterImplPort"),
+                new URL(WEBSERVICE_MOCK_ADDRESS));
+
+        final ObjectMapper mapper = new ObjectMapper();
+        final ObjectNode valuesObj = mapper.createObjectNode();
+        valuesObj.put("value1", 111);
+        valuesObj.put("value2", 222);
+        System.out.println(valuesObj);
```
Comment (dbmalkovsky) [inline modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java]: Is the here for debugging only?

File: modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java [hunk 4]
```diff
@@ -138,4 +142,171 @@ public void testWebServiceInvocationWithEndpointAddressConfigured() throws Excep
...
+                new QName("http://webservice.impl.engine.flowable.org/", "CounterImplPort"),
+                new URL(WEBSERVICE_MOCK_ADDRESS));
+
+        final ObjectMapper mapper = new ObjectMapper();
+        final ObjectNode argsObj = mapper.createObjectNode();
+        final ObjectNode valuesObj = mapper.createObjectNode();
+        valuesObj.put("arg1", 1111);
+        valuesObj.put("arg2", 2222);
+        argsObj.set("args", valuesObj);
+        System.out.println(argsObj);
```
Comment (dbmalkovsky) [inline modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java]: Again, is this debugging only?

File: modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java (hunk 3 above)
Comment (cdeneux) [reply modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java]: Yes, I will remove it

File: modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java (hunk 4 above)
Comment (cdeneux) [reply modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java]: Yes again, I will remove it

Comment (cdeneux) [reply modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java]: Fix with commit 45fb911

File: modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java (hunk 3 above)
Comment (cdeneux) [reply modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/WebServiceTaskTest.java]: Fix with commit 45fb911

File: modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/Operands.java (hunk 1 above)
Comment (cdeneux) [reply modules/flowable-cxf/src/test/java/org/flowable/engine/impl/webservice/Operands.java]: Fix with commit 45fb911

File: modules/flowable-bpmn-model/src/main/java/org/flowable/bpmn/model/JsonDataObject.java (hunk 2 above)
Comment (cdeneux) [reply modules/flowable-bpmn-model/src/main/java/org/flowable/bpmn/model/JsonDataObject.java]: Ooops, it's a relic of copy / paste.
Commit 562d5f1 fix how a value is set as Json data object

**Comment by cdeneux** [comment]
Is it possible to relaunch the test part that fails ? IMO, the failure has no link with these changes.

**Comment by filiphr** [comment]
@cdeneux, yes the failure is not linked to your changes. The Java 11 and Java 13 builds are OK. 

We will get back to you on the PR as soon as we can.

**Comment by cdeneux** [comment]
Thanks @filiphr for your feedback. Feel free to contact me for any questions, comments or additions on this PR.

**Comment by tijsrademakers** [comment]
Thanks @cdeneux for the PR. I think the changes look good. I have two points before it can be merged:

- In several place ObjectMapper objectMapper = new ObjectMapper() is done. Instead the object mapper instance should be retrieved from the engine configuration, because it's already available and instantiated there. Doing a lot of new ObjectMapper() calls is quite expensive.
- There's a TODO in the added code:
    @Override
    public Class<?> getFieldParameterTypeAt(int index) {
        // TODO
        return null;
    }

**Comment by cdeneux** [comment]
Thanks @tijsrademakers 

I'm taking into account your remark about the reuse of `ObjectMapper`.
About the TODO, I've leave it by coherence with other methods of the class `ClassStructureDefinition` (`getFieldSize`, `getFieldNameAt`, ...). What do you expect from me about this point ? 

Regards

**Comment by tijsrademakers** [comment]
@cdeneux thanks for your quick response. Ignore my comment about the TODO, because I didn't look at the whole class implementation. I understand why you added it now :-)

**Pull request merged**